### DISCUSSION TITLE
Variable-Pool #1

### DISCUSSION THREAD

**Discussion by mgrossmann** [adesutherland/CREXX#31] [General]
Hello @adesutherland ,

can you please give us a short sketch of how the assembler interface to the variable pool should look like?

/MIkeG

**Comment by adesutherland** [comment]
# NOTE - THIS IS DEPRECATED - Variable Pool will be implemented in REXX 

# Overview
In the current code for the VM we have registers which point to a variable structure (that can contain string, integers etc). These can be considered to be “anonymous variables” – the compiler will assign a register to hold a variable at compile time. 

In this way “80%” of the needs for REXX programs will be handled – but not all. Some aspects of REXX needs dynamic variable names – e.g. Some EXPOSE scenarios, INTERPRET, VALUE, and the REXXCOMM / SAA API. When these are needed the compiler needs to work in what I am calling “Pedantic” mode. In this mode variables are also given a string index in a variable pool, this index can be searched for dynamically at runtime.

# Variable Pool
Each stack frame will also include its own variable pool. This is a name/value index (via a HASH or TREE) whereby a variable can be found via the index string. 

- When a procedure exits and the stack frame is torn down, the corresponding variable pool and variables also need freeing.
- To facilitate EXPOSE, a variable pool index can be linked to a parent pool variable. This is a future requirement.
- A register can point to an anonymous variable (as implemented today in the code) or to a variable in the variable pool. This is why we don’t initiate the registers automatically because we don’t know if they should be initiated anonymously or not. 
- For the moment we are ignoring stem names – this is because STEMS may be implemented as objects (tbc)

# Instructions
(Apologies if these rename existing instructions – the name does not matter but hopefully this explains what is needed).

## Anonymous Variables

- `INIT REG` – Initiates a register with memory

- `LOAD REG, VALUE` – Puts the value into an initiated register
(Obviously, these are just one of many instructions in the same family (e.g. LOAD REG, INT, LOAD REG STRING, etc)

- `FREE REG` – Clears and deletes the variable (only really relevant for variable pool variables as anonymous ones are automatically freed)

- `MOVE REG, REG` – this moves the variable (i.e. the pointer of the target is set to that of the source – and the source pointer is set to zero – uninitialized).

- `LINK REG, REG` – I envisaged this to make both registers point to the same variable. HOWEVER I think this will lead to memory leaks and it probably a very bad idea!! Lets drop it!

This implies we should have (I think this is what we have implemented today under the “wrong” name)

- `S_LOAD REG, VALUE` - This check if the variable is initiated, and if not initiates it, and then puts the value into it.

## Variable Pool Variables
(Again, obviously, these are just one of many instructions in the same family)

- `MAPFROM REG, STRING` – This takes an existing (or makes a new) variable in the variable pool and maps it the register. Note 1/ that if the register is already initialised then the existing variable needs to be freed. 2/ This is inevitably quite a slow instruction anyway, so don’t worry, we can add safety checks without performance concerns. 

- `MAPTO REG,STRING` – This replaces the variable pool entry with the one from the register. Note 1/ The existing variable pool entry will need freeing 2/ The ownership (for memory management) of the variable moved to the variable pool. 3/ This is inevitably quite a slow instruction anyway, so don’t worry, we can add safety checks without performance concerns.

- `FREE REG` – As above (in the Anonymous Variables section) – this deletes the variable mapped to the register

- `FREE STRING` – deletes the variable with the name.

- `UNMAP REG` – this disassociates the register from the variable in the variable pool. At this point the register in not initiated anymore.

# Usage Examples

## REXX

```
a = 1
b = 2
say a + b
```

## Fast mode
```
S_LOAD R1,1
S_LOAD R2,2
IADD R0,R1,R2
SAY R0
```

## Pedantic mode
```
MAPFROM R1, “a”
MAPFROM R2, “b”
LOAD R1,1
LOAD R1,2
IADD R0,R1,R2
SAY R0
```

**Comment by adesutherland** [comment]
Note - Current plan is that the Variable Pool will be implemented in REXX Level B

**Reply by rvjansen** [reply]
Good news! I am going to ask Bob to explain a bit what the ins and outs are of the VM/370 implementation. As the linkage will now first be rxas linkage instead of native linkage, will be a good moment to design native asm and C conventions for the native interface to it. OoRexx caused a great schism in the community by letting go of the RexxSaa.h convention in order to accomodate 64bit size of strings. I am determined to mend that for cRexx (so we can use all sql, qt and other librairies Mark Hessling put together for Regina). So have to speak to Mark also. I am positive we can fix all this because we have the rxas layer.
